pkg/stream/stream.go

```diff
@@ -0,0 +1,245 @@
+
+package stream
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/flashbots/go-boost-utils/types"
+	"github.com/lthibault/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/blocknative/dreamboat/pkg/structs"
+)
+
+type Pubsub interface {
+	Publish(context.Context, string, []byte) error
+	Subscribe(context.Context, string) (chan []byte, error)
+}
+
+type Datastore interface {
+	PutPayload(context.Context, structs.PayloadKey, *structs.BlockAndTrace, time.Duration) error
+	PutHeader(ctx context.Context, hd structs.HeaderData, ttl time.Duration) error
+	CacheBlock(ctx context.Context, block *structs.CompleteBlockstruct) error
+}
+
+type StreamConfig struct {
+	Logger          log.Logger
+	ID              string
+	TTL             time.Duration
+	PubsubTopic     string // pubsub topic name for block submissions
+	StreamQueueSize int
+}
+
+type RedisStream struct {
+	Pubsub Pubsub
+
+	cacheRequests chan *structs.BlockAndTrace
+	storeRequests chan *structs.BlockAndTrace
+
+	Config StreamConfig
+	Logger log.Logger
+
+	m StreamMetrics
+}
+
+func NewRedisStream(ps Pubsub, cfg StreamConfig) *RedisStream {
+	s := RedisStream{
+		Pubsub:        ps,
+		cacheRequests: make(chan *structs.BlockAndTrace, cfg.StreamQueueSize),
+		storeRequests: make(chan *structs.BlockAndTrace, cfg.StreamQueueSize),
+		Config:        cfg,
+		Logger:        cfg.Logger.WithField("relay-service", "stream").WithField("type", "redis"),
+	}
+
+	s.initMetrics()
+
+	return &s
+}
+
+func (s *RedisStream) RunSubscriberParallel(ctx context.Context, ds Datastore, num uint) error {
+	blocks, err := s.Pubsub.Subscribe(ctx, s.Config.PubsubTopic)
+	if err != nil {
+		return err
+	}
+
+	for i := uint(0); i < num; i++ {
+		go s.RunSubscriber(ctx, ds, blocks)
+	}
+
+	return nil
+}
+
+func (s *RedisStream) RunSubscriber(ctx context.Context, ds Datastore, blocks chan []byte) error {
+	sBlock := StreamBlock{}
+
+	for rawSBlock := range blocks {
+		if err := proto.Unmarshal(rawSBlock, &sBlock); err != nil {
+			s.Logger.Warnf("fail to decode stream block: %s", err.Error())
+		}
+
+		if sBlock.Source == s.Config.ID {
+			continue
+		}
+
+		block := ToBlockAndTrace(&sBlock)
+		if sBlock.IsCache {
+			s.m.StreamRecvCounter.WithLabelValues("cache").Inc()
+			if err := s.cachePayload(ctx, ds, block); err != nil {
+				s.Logger.WithError(err).With(block).Warn("failed to cache payload")
+			}
+		} else {
+			s.m.StreamRecvCounter.WithLabelValues("store").Inc()
+			if err := s.storePayload(ctx, ds, block); err != nil {
+				s.Logger.WithError(err).With(block).Warn("failed to store payload: %s")
+			}
+		}
+	}
+
+	return ctx.Err()
+}
+
+func (s *RedisStream) RunPublisherParallel(ctx context.Context, num uint) {
+	for i := uint(0); i < num; i++ {
+		go s.RunPublisher(ctx)
+	}
+}
+
+func (s *RedisStream) RunPublisher(ctx context.Context) error {
+	for {
+		select {
```

Comment:
I was doing that quite often myself but I think Louis can calculate a proof for you from the statistical perspective it wouldn't matter. and calling multiple selects in one for may have worse performance than go randomised reads.If you like to make it prioritised I would consider just running another goroutine instead. Unless there is some particular data race case that I don't see here?

Reply:
I would recommend shipping as-is, and profiling.  I don't think we're in a pathological case here, because we're not dynamically swapping out channels, and there isn't a channel that is (almost) always in a receivable state.